### ISSUE TITLE
py==1.11.0 vulnerability 

### ISSUE THREAD

**Issue by fenchu** [pytest-dev/pytest-html#570]
Hi is it time to drop the dreaded py library, it is no longer maintained an now even freeware safety start complaining:

please see the bug: https://github.com/pytest-dev/py/issues/287

I've heard rumors about a v4 of pytest-html, so eagerly awaiting

```
20221118102134|ERROR|safety failed:
  Safety v2.3.1 is scanning for Vulnerabilities...
  Scanning dependencies in your environment:

  -> c:\dist\venvs\multitenant-fullstack-test\lib\site-packages

  Using non-commercial database
  Found and scanned 84 packages
  Timestamp 2022-11-18 10:21:34
  1 vulnerability found
  0 vulnerabilities ignored

+==============================================================================+
 VULNERABILITIES FOUND
+==============================================================================+

-> Vulnerability found in py version 1.11.0
   Vulnerability ID: 51457
   Affected spec: <=1.11.0
   ADVISORY: Py throughout 1.11.0 allows remote attackers to conduct a
   ReDoS (Regular expression Denial of Service) attack via a Subversion...
   CVE-2022-42969
   For more information, please visit https://pyup.io/v/51457/f17

 Scan was completed. 1 vulnerability was found.
```
but:
```
pip install py --upgrade
Requirement already satisfied: py in c:\dist\venvs\multitenant-fullstack-test\lib\site-packages (1.11.0)
```

**Comment by BeyondEvil** [comment]
The rumours are true @fenchu . The not-so-secret project is called `next-gen` and is "almost" done (meaning ready for its first RC).

It's taken way longer than anticipated, but in my defense, it's only been me doing the python work, and a friend of mine doing the JS.

I would **really** appreciate your feedback once we're in RC.

It's more or less a complete rewrite of the plugin but we've tried to keep it 1:1 from a feature perspective with the v3 (and older) plugin.

**Comment by BeyondEvil** [comment]
Happy to announce we've released 4.0.0rc0

**Comment by fenchu** [comment]
You get it via `pip install pytest-html==4.0.0rc`, I'll test it. Thanks :-)

**Comment by fenchu** [comment]
Do you have a guide for the new version, because my cell alterations do no longer work:
```
@pytest.mark.optionalhook
def pytest_html_results_table_header(cells):
    cells.insert(2, html.th('Description'))
    cells.insert(1, html.th('Time', class_='sortable time', col='time'))
    cells.pop()

@pytest.mark.optionalhook
def pytest_html_results_table_row(report, cells):
    if report.skipped:
        del cells[:]    # delete sckipped from list
    elif hasattr(report, 'description'):
        #log.info(f"report:{report}")
        # methods = dict((name, func) for name, func
        #                in inspect.getmembers(report))
        # for k,v in methods.items():
        #     log.info(f"{k}={v}")
        cells.insert(2, html.td(report.description))
        if hasattr(report, 'starttime'):
            cells.insert(1, html.td(str(report.starttime)[:-3], class_='col-time'))
        else:
            cells.insert(1, html.td("None", class_='col-time'))
        cells.pop()
```
`cells.pop()` and `del cells[:]` no longer works, so how do we do this now?

**Comment by BeyondEvil** [comment]
I'll get back to you a little later. I only tested the simple case.

Didn't take pop or delete into account.

**Comment by BeyondEvil** [comment]
Sorry, time ran out for me today. Ping me if I haven't gotten back to you with a response EOW.

Likely, iirc the new implementation won't allow deletion of cells. So I have to think about how we can accomplish that.

Would you mind sharing a report with me, that has the changes you're doing?

**Comment by BeyondEvil** [comment]
Still thinking about this. I have an idea of how to solve it. At least for the deletion...

What does `pop` accopmplish in this case?

**Comment by BeyondEvil** [comment]
Another thing I’m curious about, specifically for the del-case: what, if any, side-effects are there?

Does it effect the total, # of tests and duration?

What is the expectation?

**Comment by BeyondEvil** [comment]
Ping @fenchu

**Comment by prempiyush** [comment]
@fenchu Hello

I appreciate your work in this library. I can understand that the `next-gen` might take more time.. Are there any plans for fixing this security vulnerability in the meantime?

**Comment by BeyondEvil** [comment]
> @fenchu Hello
> 
> I appreciate your work in this library. I can understand that the `next-gen` might take more time.. Are there any plans for fixing this security vulnerability in the meantime?

You can start using 4.0.0rc0. The sec vuln is fixed there.

There are some known bugs with that release, most of them are fixed on the next-gen branch. Hopefully rc1 will be released some time this week.

**Comment by prempiyush** [comment]
Thank you @BeyondEvil  for a quick reply. It helped. :)

**Comment by BeyondEvil** [comment]
`4.0.0rc1` is out!

**Comment by fenchu** [comment]
will test it when Im back after easter

**Comment by timtroendle** [comment]
Is this issue solved in releases >= 4.0?

**Comment by BeyondEvil** [comment]
> Is this issue solved in releases >= 4.0?

Yes, py is no longer required.

**Comment by RonnyPfannschmidt** [comment]
Closing as it's resolved
